CompuCell3D/core/CompuCell3D/plugins/FocalPointPlasticity/FocalPointPlasticityLinkInventoryBase.h

```diff
@@ -0,0 +1,233 @@
+/*************************************************************************
+*    CompuCell - A software framework for multimodel simulations of     *
+* biocomplexity problems Copyright (C) 2003 University of Notre Dame,   *
+*                             Indiana                                   *
+*                                                                       *
+* This program is free software; IF YOU AGREE TO CITE USE OF CompuCell  *
+*  IN ALL RELATED RESEARCH PUBLICATIONS according to the terms of the   *
+*  CompuCell GNU General Public License RIDER you can redistribute it   *
+* and/or modify it under the terms of the GNU General Public License as *
+*  published by the Free Software Foundation; either version 2 of the   *
+*         License, or (at your option) any later version.               *
+*                                                                       *
+* This program is distributed in the hope that it will be useful, but   *
+*      WITHOUT ANY WARRANTY; without even the implied warranty of       *
+*  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU    *
+*             General Public License for more details.                  *
+*                                                                       *
+*  You should have received a copy of the GNU General Public License    *
+*     along with this program; if not, write to the Free Software       *
+*      Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.        *
+*************************************************************************/
+#ifndef FOCALPOINTPLASTICITYLINKINVENTORYBASE_H
+#define FOCALPOINTPLASTICITYLINKINVENTORYBASE_H
+
+#include <map>
+#include <set>
+#include <vector>
+
+#include <CompuCell3D/CC3D.h>
+
+#include "FocalPointPlasticityLinks.h"
+#include "FocalPointPlasticityDLLSpecifier.h"
+
+namespace CompuCell3D {
+
+	/**
+	Written by T.J. Sego, Ph.D.
+	*/
+
+	class Potts3D;
+	class FocalPointPlasticityTracker;
+
+	typedef std::set<FocalPointPlasticityTrackerData> FPPTrackerDataSet;
+
+	template <typename T>
+	class FPPLinkListBase : public std::vector<T*> {
+	public:
+		typedef typename std::vector<T*>::iterator FPPLinkListIterator_t;
+		virtual ~FPPLinkListBase() {}
+	};
+
+	class FOCALPOINTPLASTICITY_EXPORT FPPLinkID {
+	public:
+		long id0;
+		long id1;
+
+		FPPLinkID() : id0(0), id1(0) {};
+		FPPLinkID(long _id0, long _id1)
+		{
+			if (_id0 < _id1) {
+				id0 = _id0;
+				id1 = _id1;
+			}
+			else {
+				id1 = _id0;
+				id0 = _id1;
+			}
+		}
+		~FPPLinkID() {}
+		bool operator < (const FPPLinkID & _rhs) const { return id0 < _rhs.id0 || (id0 == _rhs.id0 && id1 < _rhs.id1); }
+		bool operator == (const FPPLinkID & _rhs) const { return id0 == _rhs.id0 && id1 == _rhs.id1; }
+		bool operator != (const FPPLinkID & _rhs) const { return !(operator==(_rhs)); }
+	};
+
+	template <class LinkType> class FPPLinkInventoryBase;
+
+	template <class LinkType>
```

Comment:
This I think these generic elements are important for a few reasons. 

First, while much of link functionality is the same for all link types, there are some particularities about the different link types (specifically external/internal vs. anchor) that change the way we interact with the data that's managed by the individual inventories. For example, we can ask an inventory that manages external or internal link what all cells are linked to a particular cell (according to the link type), which we can't ask of an anchor inventory. Likewise, we can assign the role of getting the next anchor id to the anchor inventory, but putting this function in the external or internal inventory definition is confusing. 

The same is true for the hierarchy of class definitions for the link types. Much is the same for all link types (e.g., FocalPointPlasticityLinkBase::getMaxDistance()), while some are specific to link type (e.g., FocalPointPlasticityAnchor::getAnchorPoint()). Again, we could just define a uniform interface for all link types and throw errors or something comparable when a function is used inappropriately, but this also seems to add more confusion than the complexity of this code, and I expect that it would also make inventory integration much more complicated in both the frontend and backend. 

Second, this structure should well support future dev for links. It was tempting to just assume that external and internal links will function the same, and then add some tricks to support anchors. But this assumes that the complexity of the definitions of the link types will not increase, and also that new link types will not be added in future work. The approach used in the PR is important to both of these points in that we can (hopefully) avoid a lot of supporting work for development of links by adopting a general approach now.